fix(audits): incorrect status code range

DoctorJohn opened the pull request:
This PR fixes the status code range assertion of audit check B6DC, which did not match the audit name/description.

https://github.com/graphql/graphql-http/blob/a49c45b5afa0173a65e61d87a609e1ee40142032/src/audits/server.ts#L559-L572

The name/description suggests 5xx status codes may be used. However, the audit function currently requires the status code to be between 400 and 499.

enisdenjo reviewed (changes requested):
Actually, as per the spec, the expected status code should be 400 for both content-types. Would you like to make the necessary changes?

application/json
https://graphql.github.io/graphql-over-http/draft/#sec-application-json.Examples.JSON-parsing-failure

application/graphql-response+json
https://graphql.github.io/graphql-over-http/draft/#sec-application-graphql-response-json.Examples.JSON-parsing-failure

enisdenjo commented:
Historically we allowed 4XX which is why those cases exist, but are now obsolete.

> 2. Update B6DC to allow 2xx, 400, and 5xx for compatibility with legacy servers

Note that it should allow 2XX only when accepting application/json, not the new content type.

Other than that, we're looking good. Thanks!

DoctorJohn commented:
(woops, accidentally closed this by resetting my changes to keep working on them. Going to reopen in a few minutes).

DoctorJohn committed:
fix(audits): outdated B6DC status code ranges

Relevant section: 6.4.1 application/json
https://graphql.github.io/graphql-over-http/draft/#sel-FANNLTCAACEBi8K
https://graphql.github.io/graphql-over-http/draft/#sel-FANNLVCCBCIsB7xT

DoctorJohn committed:
fix(audits): outdated BCF8 description and headers

Relevant section: 6.4.2.1.1 JSON parsing failure
https://graphql.github.io/graphql-over-http/draft/#sec-application-graphql-response-json.Examples.JSON-parsing-failure

DoctorJohn commented:
Finally found time to wrap things up. Out of the three proposed changes, I ultimately implemented numbers 2 and 3. Number 1 turned out to be unrelated to the other changes, because I mixed up document/query parsing failures and request body parsing failures.

Here's a summary of the changes:

- I updated B6DC to check the extra status codes, and clarified that the check is for the `application/json` media type
- I updated BCF8 from MAY to SHOULD to match the spec, and also clarified that the check is for the `application/json` media type

It's two separate commits, each referencing the relevant parts in the specs in the commit message body.

enisdenjo reviewed (approved):
Great work, this is looking good! 🙏
